python/pyarrow/_fs.pyx

```diff
@@ -0,0 +1,515 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# cython: language_level = 3
+
+import six
+
+from pyarrow.compat import frombytes, tobytes
+from pyarrow.includes.common cimport *
+from pyarrow.includes.libarrow cimport PyDateTime_from_TimePoint
+from pyarrow.includes.libarrow_fs cimport *
+from pyarrow.util import _stringify_path
+from pyarrow.lib import _detect_compression
+from pyarrow.lib cimport (
+    check_status,
+    NativeFile,
+    BufferedOutputStream,
+    BufferedInputStream,
+    CompressedInputStream,
+    CompressedOutputStream
+)
+
+
+cdef inline c_string _path_as_bytes(path) except *:
+    """Converts path-like objects to bytestring with posix separators
+
+    tobytes always uses utf-8, which is more or less ok, at least on Windows
+    since the C++ side then decodes from utf-8. On Unix, os.fsencode may be
+    better.
+    """
+    return tobytes(_stringify_path(path))
+
+
+cpdef enum FileType:
+    NonExistent = <int8_t> CFileType_NonExistent
+    Unknown = <int8_t> CFileType_Unknown
+    File = <int8_t> CFileType_File
+    Directory = <int8_t> CFileType_Directory
+
+
+cdef class FileStats:
```

Comment:
You should probably give this class a nice `__repr__` to make test failures more understandable ;-)